segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/hdfs/HDFSStorage.java

```diff
@@ -196,85 +323,309 @@ public boolean supportsTruncation() {
+                }
+                //Ensure that file exists
+                findStatusForSegment(streamSegmentName, true);
+                return HDFSSegmentHandle.write(streamSegmentName);
+            } catch (IOException e) {
+                throw HDFSExceptionHelpers.convertException(streamSegmentName, e);
+            }
+        } while (fencedCount < MAX_ATTEMPT_COUNT);
+        LoggerHelpers.traceLeave(log, "openWrite", traceId, epoch);
+        throw new StorageNotPrimaryException("Not able to fence out other writers.");
     }
 
     @Override
-    @SneakyThrows(StreamSegmentException.class)
-    public boolean exists(String streamSegmentName) {
-        return call(new ExistsOperation(streamSegmentName, this.context));
-    }
+    public SegmentProperties create(String streamSegmentName) throws StreamSegmentException {
+        // Creates a file with the lowest possible epoch (0).
+        // There is a possible race during create where more than one segmentstore may be trying to create a streamsegment.
+        // If one create is delayed till long time, it is possible that other segmentstore will be able to create the file with
+        // epoch (0) and then rename it to its owned version (segment_<epoch>).
+        //
+        // To fix this, the create code checks whether a file with higher epoch exists.
+        // If it does, it tries to remove the created file, and throws SegmentExistsException.
+
+        long traceId = LoggerHelpers.traceEnter(log, "create", streamSegmentName);
+        // Create the segment using our own epoch.
+        FileStatus[] status = null;
+        try {
+            status = findAllRaw(streamSegmentName);
+        } catch (IOException e) {
+            throw HDFSExceptionHelpers.convertException(streamSegmentName, e);
+        }
+        if (status != null && status.length > 0) {
+            // Segment already exists; don't bother with anything else.
+            throw HDFSExceptionHelpers.convertException(streamSegmentName, HDFSExceptionHelpers.segmentExistsException(streamSegmentName));
+        }
+
+        // Create the file for the segment with epoch 0.
+        Path fullPath = getFilePath(streamSegmentName, 0);
+        try {
+            // Create the file, and then immediately close the returned OutputStream, so that HDFS may properly create the file.
+            this.fileSystem.create(fullPath, READWRITE_PERMISSION, false, 0, this.config.getReplication(),
+                    this.config.getBlockSize(), null).close();
+            log.debug("Created '{}'.", fullPath);
+        } catch (IOException e) {
+            throw HDFSExceptionHelpers.convertException(streamSegmentName, e);
+        }
 
+        //If there is a race during creation, delete the file with epoch 0 and throw exception.
+        try {
+            status = findAllRaw(streamSegmentName);
+            if (status != null && status.length > 1) {
+                this.fileSystem.delete(fullPath, true);
```

Comment:
There must be some property that we can use to justify why deleting this file is safe. I suspect that:
- The file is guaranteed to be empty because it has just been created
- No other server will use this file in any way

Independent of whether this is right or wrong, please expand the the comment to explain why it is safe to delete it.